src/Symfony/Component/EventDispatcher/CompiledEventDispatcher.php

```diff
@@ -0,0 +1,241 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\EventDispatcher;
+
+use Symfony\Component\DependencyInjection\IntrospectableContainerInterface;
+
+/**
+ * A performance optimized container aware event dispatcher.
+ *
+ * This version of the event dispatcher contains the following optimizations
+ * in comparison to the Symfony event dispatcher component:
+ *
+ * <dl>
+ *   <dt>Faster instantiation of the event dispatcher service</dt>
+ *   <dd>
+ *     Instead of calling <code>addSubscriberService</code> once for each
+ *     subscriber, a precompiled array of listener definitions is passed
+ *     directly to the constructor. This is faster by an order of magnitude.
+ *     The listeners are collected and prepared using a compiler
+ *     pass.
+ *   </dd>
+ *   <dt>Lazy instantiation of listeners</dt>
+ *   <dd>
+ *     Services are only retrieved from the container just before invocation.
+ *     Especially when dispatching the KernelEvents::REQUEST event, this leads
+ *     to a more timely invocation of the first listener. Overall dispatch
+ *     runtime is not affected by this change though.
+ *   </dd>
+ * </dl>
+ */
+class CompiledEventDispatcher implements EventDispatcherInterface
+{
+    /**
+     * The service container.
+     *
+     * @var \Symfony\Component\DependencyInjection\IntrospectableContainerInterface
+     */
+    private $container;
+
+    /**
+     * Listener definitions.
+     *
+     * A nested array of listener definitions keyed by event name and priority.
+     * A listener definition is an associative array with one of the following key
+     * value pairs:
+     * - callable: A callable listener
+     * - service: An array of the form array(service id, method)
+     *
+     * A service entry will be resolved to a callable only just before its
+     * invocation.
+     *
+     * @var array
+     */
+    private $listeners;
+
+    /**
+     * Whether listeners need to be sorted prior to dispatch, keyed by event name.
+     *
+     * @var array
+     */
+    private $unsorted = array();
+
+    /**
+     * Constructs a container aware event dispatcher.
+     *
+     * @param \Symfony\Component\EventDispatcher\IntrospectableContainerInterface $container
+     *   The service container.
+     * @param array $listeners
+     *   A nested array of listener definitions keyed by event name and priority.
+     *   The array is expected to be ordered by priority. A listener definition is
```

Comment:
Several listeners can have the same priority, which means your array representation here is broken. You should support having multiple definitions by priority (and ``addListener`` is currently expecting a different structure for this array)